_posts/2023-01-06-Algorithms-Graphs-BFS-DFS.md

```diff
@@ -0,0 +1,338 @@
+| **Cyclce detection**                                 | Commonly used for cycle detection                            | Not very efficient for finding a cycle in the graph.        |
+| **Complexity**                                       | O(V+ E)                                                      | O(V+ E)                                                      |
+| **Practice**                                         | - https://leetcode.com/problems/shortest-path-in-a-grid-with-obstacles-elimination/description/ | https://leetcode.com/problems/minimum-genetic-mutation/solution/    https://leetcode.com/problems/nearest-exit-from-entrance-in-maze/ |
+
+### Visualization
+
+#### DFS
+
+As we said, the DFS is going branch by branch, is going deep as much as possible than it turns back and starts investigating a new branch. Having this in mind let's look at what could be the order of visiting nodes if we would start at node with id equal to 3:
+
+<p align="center">
+<img src="https://raw.githubusercontent.com/swiftingio/blog/%2357-Algorithms-DFS-BFS/DFS.svg" width="643">
+</p>
+
+Algorithm starts with the selection of some first branch and goes to the end (end means that encounters node with no other children, or all the children are already visited):
+
+Node 5 has no other children so we come back to the place when it can check another branch (it this case it will be a node 3)
+
+And then algorithm starts to vist the new unexplored branch, so it goes to node one and then it finishes the job.
+
+#### BFS
+
+The visualization shows that we go level by level and I think this is the best way how we can imaging the algorithm .
+
+<p align="center">
+<img src="https://raw.githubusercontent.com/swiftingio/blog/%2357-Algorithms-DFS-BFS/BFS.svg" width="643">
+</p>
+
+If we start with node id = 3 (first layer), then we visit 1 and 4 (second layer) then 2 (third layer) and last will be 5 (fourth layer).
+
+### Impmenenation
+
+Having this knowledge let's implement DFS and BFS with Swift. Additionally let's take a look at some additional topics which are very interesting al well:
+
+- Finding the shortest path with unweighted graph
+- Finding cycle in an undirected graph
+- Finding cycle in a directed graph
+
+#### DFS
+
+```swift
+var visited: Set<Int> = []
+
+func dfs(_ root: TreeNode?)  {
+
+	guard let root = root else {
+		return
+	}
+
+	visited.insert(root.val)
+
+	for child in root.children {
+		if !visited.contains(child.val) {
+			dfs(child)
+		}
+	}
+}
+```
+
+#### BFS
+
+```swift
+func bfs(_ start: Node) {
+
+	var queue: [Node] = []
+	var visited: Set<Node> = []
+
+	queue.append(start)
+
+	while !queue.isEmpty { // maybe then this while is not needed anymore (worth to check)
+  	for item in queue {
```

Comment:
Ta pętla wydaje się być niepotrzebna i poza tym może się rozjebać, bo linijce 156 jest modyfikowana kolejka po której iteruje ta pętla

Reply:
To powinno zostać zaimplementowane jak [poniżej](https://github.com/swiftingio/swiftingio.github.io/pull/10/files#r1103834825)